net-tools/netwatch/src/udp.rs

```diff
@@ -51,97 +65,429 @@ impl UdpSocket {
         Self::bind_raw(addr)
     }
 
+    /// Is the socket broken and needs a rebind?
+    pub fn is_broken(&self) -> bool {
+        self.is_broken.load(std::sync::atomic::Ordering::SeqCst)
+    }
+
+    /// Marks this socket as needing a rebind
+    pub fn mark_broken(&self) {
```

Comment:
this is all in progress, don't worry about the api too much for now

Reply:
I would also prefer to go with setting the `Option` to `None` since it makes this less vulnerable to toctou races. 

Maybe there's something that prevents setting `is_broken` in a non synchronized way with `rebind` (which is called after `is_broken()`) in the higher level calls, but since this is exposed as part of the public i think it would be better to have rebinding (the use part) and is_broken (the check part) as a single locked entity.

We currently use the option only for drop but we could simply do the fancy drop routine when there's actually a socket

Reply:
Going a bit further we could have a enum like
```rust
enum InnerSock {
	Active(tokio::net::UddSocket),
	Rebind(std::net::SocketAddr),
}
```

should use a teeny tiny less memory, does not require storing the address when not needed, no need to think about races about whether being broken and actually rebinding are being updated in some incorrect way and uses a single lock

Reply:
> Maybe there's something that prevents setting `is_broken` in a non synchronized way with `rebind` (which is called after `is_broken()`) in the higher level calls, but since this is exposed as part of the public i think it would be better to have rebinding (the use part) and is_broken (the check part) as a single locked entity.

I understand the desire, but `mark_broken` needs to work without taking a write lock, so it needs to be seperated into an atomic, or its own lock